sling/nlp/parser/trainer/transition-generator.h

```diff
@@ -0,0 +1,37 @@
+#ifndef SLING_NLP_PARSER_TRAINER_TRANSITION_GENERATOR_H_
+#define SLING_NLP_PARSER_TRAINER_TRANSITION_GENERATOR_H_
+
+#include <functional>
+
+#include "sling/nlp/document/document.h"
+#include "sling/nlp/parser/parser-action.h"
+
+namespace sling {                  
+namespace nlp {
+  
+// Generates transition sequences for [begin, end) token range in 'document',
+// calling 'callback' for every transition, and 'final_callback' at the end.
+void Generate(const Document &document,
+              int begin, int end,
+              std::function<void(const ParserAction &)> callback,
+              std::function<void()> final_callback);     
```

Comment:
What is the purpose of the "final" callback?

Reply:
To call a function after all the transitions have been generated.
For example, 'callback' can store transitions into a vector, and 'final_callback' can dispatch the vector to the trainer or a file writer.